AppCenter/AppCenter/Internals/Storage/MSDBStorage.m

```diff
@@ -264,100 +275,172 @@ + (void)enableAutoVacuumInOpenedDatabase:(void *)db {
-  } else if (result == SQLITE_FULL) {
-    MSLogDebug([MSAppCenter logTag], @"Query failed with error: %d - %@", result, [NSString stringWithUTF8String:errMsg]);
-  } else if (result != SQLITE_OK) {
-    MSLogError([MSAppCenter logTag], @"Query \"%@\" failed with error: %d - %@", query, result, [NSString stringWithUTF8String:errMsg]);
++ (int)executeNonSelectionQuery:(NSString *)query inOpenedDatabase:(void *)db withValues:(nullable NSArray *)values {
+  return
+      [MSDBStorage executeQuery:query
+               inOpenedDatabase:db
+                     withValues:values
+                     usingBlock:^(void *statement) {
+                       int stepResult = sqlite3_step(statement);
+                         if (stepResult == SQLITE_DONE) {
+                            return SQLITE_OK;
+                         }
+                         NSString *errorMessage = [NSString stringWithUTF8String:sqlite3_errmsg(db)];
+                         if (stepResult == SQLITE_CORRUPT || stepResult == SQLITE_NOTADB) {
+                           MSLogError([MSAppCenter logTag], @"A database file is corrupted, result=%d\n\t%@", stepResult, errorMessage);
+                         } else if (stepResult == SQLITE_FULL) {
+                           MSLogDebug([MSAppCenter logTag], @"Query failed with error: %d\n\t%@", stepResult, errorMessage);
+                         } else {
+                           MSLogError([MSAppCenter logTag], @"Could not execute the statement, result=%d\n\t%@", stepResult, errorMessage);
+                         }
+                         return stepResult;
+                     }];
+}
+
++ (int)executeQuery:(NSString *)query
+    inOpenedDatabase:(void *)db
+          withValues:(nullable NSArray *)values
+          usingBlock:(MSDBStorageQueryBlock)block {
+  sqlite3_stmt *statement = NULL;
+  int result = sqlite3_prepare_v2(db, [query UTF8String], -1, &statement, NULL);
+  if (result != SQLITE_OK) {
+    NSString *errorMessage = [NSString stringWithUTF8String:sqlite3_errmsg(db)];
+    MSLogError([MSAppCenter logTag], @"Failed to prepare SQLite statement, result=%d\n\t%@", result, errorMessage);
+    return result;
   }
-  if (errMsg) {
-    sqlite3_free(errMsg);
+  result = [MSDBStorage bindStatement:statement inOpenedDatabase:db withValues:values];
+  if (result == SQLITE_OK) {
+    result = block(statement);
+  }
+  int finalizeResult = sqlite3_finalize(statement);
+  if (finalizeResult != SQLITE_OK) {
+    NSString *errorMessage = [NSString stringWithUTF8String:sqlite3_errmsg(db)];
+    MSLogError([MSAppCenter logTag], @"Filed to finalize SQLite statement, result=%d\n\t%@", finalizeResult, errorMessage);
   }
   return result;
 }
 
-- (NSArray<NSArray *> *)executeSelectionQuery:(NSString *)query {
++ (int)bindStatement:(sqlite3_stmt *)query inOpenedDatabase:(void *)db withValues:(nullable NSArray *)values {
```

Comment:
`bindStatement:inOpenedDatabase:toValues` would be more correct I think because we bind the statements to the values not "with".